File: termin/chronosquad/core/object_time.py
"""Object time management - broken time and time modifiers."""

from __future__ import annotations
from .event_line import EventCard, TimeDirection
from .timeline import GAME_FREQUENCY


class TimeModifier(EventCard[None]):
    """Base class for time modifiers (freeze, haste, etc.)."""

    def time_offset(self, step: int) -> int:
        """Time offset at given step."""
        return 0

    def time_offset_on_finish(self) -> int:
        """Final time offset when modifier ends."""
        return 0

    def multiplier(self) -> float:
        """Time speed multiplier."""
        return 1.0

    def copy(self) -> TimeModifier:
        return TimeModifier(self.start_step, self.finish_step)


class TimeFreeze(TimeModifier):
    """Freezes object's local time."""

    def __init__(self, start_step: int, finish_step: int):
        super().__init__(start_step, finish_step)
        self._mul = -1

    def time_offset(self, step: int) -> int:
        from_start = step - self.start_step
        return from_start * self._mul

    def time_offset_on_finish(self) -> int:
        from_start = self.finish_step - self.start_step
        return from_start * self._mul

    def multiplier(self) -> float:
        return float(self._mul)

    def copy(self) -> TimeFreeze:
        return TimeFreeze(self.start_step, self.finish_step)


class TimeHaste(TimeModifier):
    """Speeds up object's local time."""

    def __init__(self, start_step: int, finish_step: int, speed: float = 2.0):
        super().__init__(start_step, finish_step)
        self._speed = speed

    def time_offset(self, step: int) -> int:
        from_start = step - self.start_step
        return int(from_start * (self._speed - 1))

    def time_offset_on_finish(self) -> int:
        from_start = self.finish_step - self.start_step
        return int(from_start * (self._speed - 1))

    def multiplier(self) -> float:
        return self._speed

    def copy(self) -> TimeHaste:
        return TimeHaste(self.start_step, self.finish_step, self._speed)


class ObjectTime:
    """
    Manages object's personal time relative to timeline.

    Timeline time (global) -> Broken time (personal) -> Local time (with modifiers)

    "Broken time" always increases even when object travels back in time.
    This allows objects to have continuous personal history.
    """

    def __init__(self):
        self._timeline_zero: int = 0
        self._broken_timeline_zero: int = 0
        self._non_finished_offset: int = 0
        self._finished_offset: int = 0
        self._is_reversed: bool = False
        self._modifiers: list[TimeModifier] = []
        self._active_modifiers: list[TimeModifier] = []

    def copy(self) -> ObjectTime:
        ot = ObjectTime()
        ot._timeline_zero = self._timeline_zero
        ot._broken_timeline_zero = self._broken_timeline_zero
        ot._non_finished_offset = self._non_finished_offset
        ot._finished_offset = self._finished_offset
        ot._is_reversed = self._is_reversed
        ot._modifiers = [m.copy() for m in self._modifiers]
        ot._active_modifiers = []
        return ot

    @property
    def timeline_zero(self) -> int:
        return self._timeline_zero

    @property
    def broken_timeline_zero(self) -> int:
        return self._broken_timeline_zero

    @property
    def is_reversed(self) -> bool:
        return self._is_reversed

    @property
    def offset(self) -> int:
        return self._non_finished_offset + self._finished_offset

    def set_reference_point(
        self,
        new_timeline_zero: int,
        new_direction_is_reversed: bool,
        offset: int = 0,
    ) -> None:
        """Set new time reference point (when object reverses direction)."""
        dist = new_timeline_zero - self._timeline_zero - offset
        broken_dist = -dist if self._is_reversed else dist
        self._broken_timeline_zero += broken_dist
        self._timeline_zero = new_timeline_zero
        self._is_reversed = new_direction_is_reversed

    def timeline_to_broken(self, timeline_step: int) -> int:
        """Convert timeline step to broken (personal) step."""
        dist = timeline_step - self._timeline_zero
        broken_dist = -dist if self._is_reversed else dist
        return broken_dist + self._broken_timeline_zero

    def broken_to_timeline(self, broken_step: int) -> int:
        """Convert broken step back to timeline step."""
        broken_dist = broken_step - self._broken_timeline_zero
        dist = -broken_dist if self._is_reversed else broken_dist
        return self._timeline_zero + dist

    def timeline_to_local(self, timeline_step: int) -> int:
        """Convert timeline step to local time (with modifiers)."""
        broken_step = self.timeline_to_broken(timeline_step)
        return self._non_finished_offset + self._finished_offset + broken_step

    def timeline_to_local_time(self, timeline_time: float) -> float:
        """Convert timeline time (seconds) to local time."""
        step = int(timeline_time * GAME_FREQUENCY)
        local_step = self.timeline_to_local(step)
        return local_step / GAME_FREQUENCY

    def add_modifier(self, modifier: TimeModifier) -> None:
        """Add a time modifier."""
        self._modifiers.append(modifier)

    def promote(self, timeline_step: int) -> None:
        """Update modifiers for current timeline step."""
        broken_step = self.timeline_to_broken(timeline_step)

        new_active: list[TimeModifier] = []
        for mod in self._modifiers:
            if mod.is_active_at(broken_step):
                new_active.append(mod)

        # Handle modifiers that just ended (forward) or just started (backward)
        old_set = set(id(m) for m in self._active_modifiers)
        new_set = set(id(m) for m in new_active)

        # Modifiers that were active but aren't anymore
        for mod in self._active_modifiers:
            if id(mod) not in new_set:
                # Modifier ended - add its final offset
                self._finished_offset += mod.time_offset_on_finish()

        # Modifiers that weren't active but are now (going backward)
        for mod in new_active:
            if id(mod) not in old_set:
                if broken_step < mod.start_step:
                    # We're going backward into a modifier
                    self._finished_offset -= mod.time_offset_on_finish()

        self._active_modifiers = new_active
        self._update_offset(broken_step)

    def _update_offset(self, broken_step: int) -> None:
        """Update non-finished offset from active modifiers."""
        result = 0
        for mod in self._active_modifiers:
            result += mod.time_offset(broken_step)
        self._non_finished_offset = result

    def current_time_multiplier(self) -> float:
        """Get current time multiplier from all active modifiers."""
        result = 1.0
        for mod in self._active_modifiers:
            result *= mod.multiplier()
        return result

    def drop_to_current(self) -> None:
        """Drop future modifiers."""
        # Keep only modifiers that have started
        self._modifiers = [
            m for m in self._modifiers
            if m.start_step <= self._broken_timeline_zero
        ]

    def info(self) -> str:
        lines = [
            f"timeline_zero: {self._timeline_zero}",
            f"broken_timeline_zero: {self._broken_timeline_zero}",
            f"non_finished_offset: {self._non_finished_offset}",
            f"finished_offset: {self._finished_offset}",
            f"is_reversed: {self._is_reversed}",
            f"modifiers: {len(self._modifiers)}",
            f"active_modifiers: {len(self._active_modifiers)}",
        ]
        return "\n".join(lines)

File: termin/chronosquad/ChronoCoreC#/Collections/ObjectTime.cs
using System.Collections.Generic;
using System;

#if UNITY_64
using UnityEngine;
#endif

public class ObjectTime
{
    long _timeline_zero = 0;
    long _broken_timeline_zero = 0;
    long _non_finished_offset = 0;
    long _finished_offset = 0;
    bool _is_reversed = false;
    MultipleActionList<TimeModifier> _modifiers = new MultipleActionList<TimeModifier>(true);

    public ObjectTime() { }

    public MultipleActionList<TimeModifier> Modifiers => _modifiers;

    public Dictionary<string, object> ToTrent()
    {
        Dictionary<string, object> dict = new Dictionary<string, object>();
        dict["timeline_zero"] = _timeline_zero;
        dict["broken_timeline_zero"] = _broken_timeline_zero;
        dict["non_finished_offset"] = _non_finished_offset;
        dict["finished_offset"] = _finished_offset;
        dict["is_reversed"] = _is_reversed;
        dict["modifiers"] = _modifiers.ToTrent();
        return dict;
    }

    public void FromTrent(Dictionary<string, object> dict)
    {
        _timeline_zero = (long)dict["timeline_zero"];
        _broken_timeline_zero = (long)dict["broken_timeline_zero"];
        _non_finished_offset = (long)dict["non_finished_offset"];
        _finished_offset = (long)dict["finished_offset"];
        _is_reversed = (bool)dict["is_reversed"];
        _modifiers.FromTrent((Dictionary<string, object>)dict["modifiers"]);
    }

    public bool IsEqual(ObjectTime other)
    {
        // var this_modifiers = _modifiers.AsList();
        // var other_modifiers = other._modifiers.AsList();

        if (_modifiers.Count != other._modifiers.Count)
        {
            return false;
        }

        // for (int i = 0; i < this_modifiers.Count; ++i)
        // {
        // 	if (!this_modifiers[i].IsEqual(other_modifiers[i]))
        // 	{
        // 		return false;
        // 	}
        // }

        return true;
    }

    public ObjectTime(ObjectTime other)
    {
        _timeline_zero = other._timeline_zero;
        _broken_timeline_zero = other._broken_timeline_zero;
        _non_finished_offset = other._non_finished_offset;
        _finished_offset = other._finished_offset;
        _modifiers = new MultipleActionList<TimeModifier>(other._modifiers);
        _is_reversed = other._is_reversed;
    }

    public void SetReferencePoint(
        long new_timeline_zero,
        bool newdirection_is_reversed,
        long offset = 0
    )
    {
        long dist = new_timeline_zero - _timeline_zero - offset;
        long broken_dist = _is_reversed ? -dist : dist;
        _broken_timeline_zero += broken_dist;
        _timeline_zero = new_timeline_zero;
        _is_reversed = newdirection_is_reversed;
    }

    public long TimelineZero => _timeline_zero;
    public long BrokenTimelineZero => _broken_timeline_zero;
    public long NonFinishedOffset => _non_finished_offset;
    public long FinishedOffset => _finished_offset;
    public bool IsReversed => _is_reversed;

    public long Offset
    {
        get { return _non_finished_offset + _finished_offset; }
    }

    // b (broken)
    // |   /
    // |  /
    // | .
    // |  \
    // |   \ (timeline_zero)
    // |----.
    // |   / | (broken_timeline_zero)
    // |  /  |
    // | /   |
    // |/    |
    // ------------------------------ t (timeline)
    public long TimelineToBroken(long nonbroken_timeline_step)
    {
        long dist = nonbroken_timeline_step - _timeline_zero;
        long broken_dist = _is_reversed ? -dist : dist;
        long broken = broken_dist + _broken_timeline_zero;
        return broken;
    }

    public long ToBroken(long s)
    {
        return TimelineToBroken(s);
    }

    public void Clean()
    {
        _modifiers.Clean();
    }

    public float ToBroken_Time(float nonbroken_timeline_time)
    {
        float dist = nonbroken_timeline_time - TimelineZero_Time();
        float broken_dist = _is_reversed ? -dist : dist;
        float broken = broken_dist + BrokenTimelineZero_Time();
        return broken;
    }

    public float TimelineZero_Time()
    {
        return (float)_timeline_zero / Utility.GAME_GLOBAL_FREQUENCY;
    }

    public float BrokenTimelineZero_Time()
    {
        return (float)_broken_timeline_zero / Utility.GAME_GLOBAL_FREQUENCY;
    }

    public long BrokenToTimeline(long broken)
    {
        long broken_dist = broken - _broken_timeline_zero;
        long dist = _is_reversed ? -broken_dist : broken_dist;
        long nonbroken_timeline_step = _timeline_zero + dist;
        return nonbroken_timeline_step;
    }

    public long BrokenPointInTimeline()
    {
        return _broken_timeline_zero;
    }

    public string info()
    {
        string text = "";
        long _timeline_zero = TimelineZero;
        long _broken_timeline_zero = BrokenTimelineZero;
        long _non_finished_offset = NonFinishedOffset;
        long _finished_offset = FinishedOffset;
        bool _is_reversed = IsReversed;

        text += "timeline_zero: " + _timeline_zero + "\n";
        text += "broken_timeline_zero: " + _broken_timeline_zero + "\n";
        text += "non_finished_offset: " + _non_finished_offset + "\n";
        text += "finished_offset: " + _finished_offset + "\n";
        text += "is_reversed: " + _is_reversed + "\n";

        // text += "sorted by start: \n";
        // foreach (var modifier in _modifiers.AsList())
        // {
        // 	text += modifier.info() + "\n";
        // }

        // text += "sorted by finish: \n";
        // foreach (var modifier in _modifiers.AsListByFinish())
        // {
        // 	text += modifier.info() + "\n";
        // }

        text += "by_start pos: " + _modifiers.ByStartIteratorPosition() + "\n";
        text += "by_finish pos: " + _modifiers.ByFinishIteratorPosition() + "\n";

        return text;
    }

    public void AddModifier(TimeModifier modifier) => _modifiers.Add(modifier);

    public MyList<TimeModifier> ActiveModifiers() => _modifiers.ActiveStates();

    public long TimelineToLocal(long step)
    {
        var broken_step = TimelineToBroken(step);
        return BrokenToLocal(broken_step);
    }

    public long BrokenToLocal(long broken_step)
    {
        return _non_finished_offset + _finished_offset + broken_step;
    }

    // public long LocalToTimeline(long step)
    // {
    // 	var broken_step = step - _non_finished_offset - _finished_offset;
    // 	var tlstep = BrokenToTimeline(broken_step);
    // 	Debug.Log("LocalToTimeline step: " + step + " broken_step: " + broken_step + " tlstep: " + tlstep);
    // 	return tlstep;
    // }

    public float NonFinishedOffset_Time()
    {
        return _non_finished_offset / Utility.GAME_GLOBAL_FREQUENCY;
    }

    public float FinishedOffset_Time()
    {
        return _finished_offset / Utility.GAME_GLOBAL_FREQUENCY;
    }

    public float TimelineToLocal_Time(float time)
    {
        var broken_time = ToBroken_Time(time);
        return NonFinishedOffset_Time() + FinishedOffset_Time() + broken_time;
    }

    public float TimelineToLocal_Seconds(float t)
    {
        return TimelineToLocal((long)(t * Utility.GAME_GLOBAL_FREQUENCY))
            / Utility.GAME_GLOBAL_FREQUENCY;
    }

    public void UpdateOffset(long step)
    {
        long result = 0;
        foreach (var modifier in ActiveModifiers())
        {
            if (modifier == null)
                continue;

            var modoffset = modifier.TimeOffset(step);
            result += modoffset;
        }
        _non_finished_offset = result;
    }

    public Step TimelineToLocal(Step step)
    {
        return new Step(TimelineToLocal(step.step), step.direction);
    }

    public void Promote(long timeline_step)
    {
        long broken_step = ToBroken(timeline_step);
        //Step broken = new Step(broken_step, timeline_step.direction);

        MyList<TimeModifier> added;
        MyList<TimeModifier> goned;
        TimeDirection direction;
        _modifiers.Promote(broken_step, out added, out goned, out direction);

        if (direction == TimeDirection.Forward)
        {
            foreach (var modifier in goned)
            {
                if (modifier == null)
                    continue;

                _finished_offset += modifier.TimeOffsetOnFinish();
            }
        }
        else
        {
            foreach (var modifier in added)
            {
                if (modifier == null)
                    continue;

                _finished_offset -= modifier.TimeOffsetOnFinish();
            }
        }

        UpdateOffset(broken_step);
    }

    public void DropToCurrentState()
    {
        _modifiers.DropToCurrentState();
    }

    public void DropToCurrentStateInverted()
    {
        _modifiers.DropToCurrentStateInverted();
    }

    public int CountOfModifiers()
    {
        return _modifiers.Count;
    }

    public float CurrentTimeMultiplier()
    {
        float result = 1.0f;
        foreach (var modifier in ActiveModifiers())
        {
            if (modifier == null)
                continue;

            result *= modifier.Multiplier();
        }
        return result;
    }

    public void AddBrokenOffsetToLocalTime(long offset)
    {
        // _finished_offset += offset;
    }
}
